### PR TITLE
Add compatibility with elasticsearch  >=  8.4

### PR DESCRIPTION
**Pull request by sidz**
| Q             | A
| ------------- | ---
| Branch?       | main
| Tickets       | https://github.com/api-platform/core/issues/5430
| License       | MIT

This PR adds elasticsearch 8 compatibility.

I've choose 8.4 as ClientInterface was introduced since 8.4 version and Client is marked as final.

### CODE REVIEW

File: tests/Elasticsearch/Metadata/Resource/Factory/ElasticsearchProviderResourceMetadataCollectionFactoryTest.php [hunk 1]
```diff
@@ -80,10 +82,12 @@ public function testCreate(?bool $elasticsearchFlag, int $expectedCatCallCount,
...
             ]]);
         } else {
-            $catNamespace->indices(['index' => 'foo'])->willThrow(new Missing404Exception());
+            $catNamespace->indices(['index' => 'foo'])->willThrow(new ClientResponseException());
         }
 
-        $client = $this->prophesize(Client::class);
+        $clientClass = class_exists(\Elasticsearch\Client::class) ? \Elasticsearch\Client::class : \Elastic\Elasticsearch\ClientInterface::class;
+
+        $client = $this->prophesize($clientClass);
```
Comment (sidz) [inline tests/Elasticsearch/Metadata/Resource/Factory/ElasticsearchProviderResourceMetadataCollectionFactoryTest.php:90]: `Elastic\Elasticsearch\ClientInterface` doesn't have `cat` method defined inside of it and I don't have a clue how to mock method which doesn't exist with a help of **prophecy**. On other hand I couldn't mock `Elastic\Elasticsearch\Client` as it is marked as **final**. 

I see only one way here: write decorator for a ES Client on Api-Platform side

File: tests/Behat/ElasticsearchContext.php [hunk 2]
```diff
@@ -88,15 +89,17 @@ private function deleteIndexes(): void
...
-            $index = $file->getBasename('.json');
-            if (!$indexClient->exists(['index' => $index])) {
-                continue;
-            }
+            $indexes[] = $file->getBasename('.json');
+        }
 
-            $indexClient->delete(['index' => $index]);
+        if ($indexes !== []) {
+            $this->client->indices()->delete([
```
Comment (sidz) [inline tests/Behat/ElasticsearchContext.php:99]: remove all indexes within a single http request instead of multiple HTTP requests

**Comment by soyuka** [comment]
this looks really awesome, I'll review on friday / this week-end can't for now but I've got your messages!

File: src/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php [hunk 3]
```diff
@@ -718,6 +718,13 @@ private function registerElasticsearchConfiguration(ContainerBuilder $container,
             return;
         }
 
+        $clientClass = class_exists(\Elasticsearch\Client::class) ? \Elasticsearch\Client::class : \Elastic\Elasticsearch\Client::class;
+
+        $clientDefinition = new Definition($clientClass);
+        $clientDefinition->setPublic(false);
```
Comment (soyuka) [inline src/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php:724]: you don't need that or do you?

File: tests/Elasticsearch/Metadata/Resource/Factory/ElasticsearchProviderResourceMetadataCollectionFactoryTest.php (hunk 1 above)
Comment (soyuka) [reply tests/Elasticsearch/Metadata/Resource/Factory/ElasticsearchProviderResourceMetadataCollectionFactoryTest.php:90]: We should avoid changing existing tests, duplicate the tests you need, don't test the `cat` with es > 8

File: src/Elasticsearch/Metadata/Document/Factory/CatDocumentMetadataFactory.php [hunk 4]
```diff
@@ -30,7 +32,7 @@
  */
 final class CatDocumentMetadataFactory implements DocumentMetadataFactoryInterface
 {
-    public function __construct(private readonly Client $client, private readonly ResourceMetadataCollectionFactoryInterface $resourceMetadataFactory, private readonly ?DocumentMetadataFactoryInterface $decorated = null)
+    public function __construct(private readonly Client|ClientInterface $client, private readonly ResourceMetadataCollectionFactoryInterface $resourceMetadataFactory, private readonly ?DocumentMetadataFactoryInterface $decorated = null)
```
Comment (soyuka) [inline src/Elasticsearch/Metadata/Document/Factory/CatDocumentMetadataFactory.php:35]: should we deprecate the old client? how can we align on Elasticsearch updates in the future? The timeline on this is to release this as part of the 3.2 release which will be in september.

**Review by soyuka — COMMENTED** [review]
Nice job so far, some few leftovers.

File: src/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php (hunk 3 above)
Comment (sidz) [reply src/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php:724]: Just did it as it was before: https://github.com/api-platform/core/pull/5503/files#diff-e55025391dae392739748b2ff664550ac2bb4b800005b9f22da70fb43a8eec4bL8

**Comment by turegjorup** [comment]
Would love to see this ready for `3.2` - any way to help get this PR going again?

File: src/Elasticsearch/Metadata/Document/Factory/CatDocumentMetadataFactory.php (hunk 4 above)
Comment (turegjorup) [reply src/Elasticsearch/Metadata/Document/Factory/CatDocumentMetadataFactory.php:35]: I would consider letting the versioning for a `api-platform/elasticsearch` match the `elasticsearch/elastichsearch` [versioning](https://github.com/elastic/elasticsearch-php#versioning) even though it means starting the versioning af `7.*`

**Comment by soyuka** [comment]
wow stale bot this wasn't supposed to be close :D sorry!

**Comment by ToshY** [comment]
@soyuka 

1. Should #5430 not be re-opened until this is merged?
2. The milestone for `3.2` is set for [`September 1, 2023`](https://github.com/api-platform/core/milestone/15). Will this feature be included?

**Comment by soyuka** [comment]
yes someone needs to rebase and finish this I think that we should just remove the TypeHint for now within the providers so keep them as flexible as possible for the feature (just add a phpdoc).

**Comment by sidz** [comment]
I'll do it

**Comment by sidz** [comment]
ready for re-review @soyuka

**Comment by soyuka** [comment]
I'll continue this at #5795

**Pull request closed without merging**